File: pset1/credit/credit.c
#include <cs50.h>
#include <stdbool.h>
#include <stdio.h>


int sum_digits(int number)
{
    if (number > 9)
    {
        return number % 10 + 1;
    }
    return number;
}


bool check_luhn(long cc_number)
{
    long number = cc_number;
    long remainder = 0;
    int even_pos_sum = 0;
    int odd_pos_sum = 0;

    int idx = 1;

    while (number != 0)
    {
        remainder = number % 10;
        if (idx % 2 == 0)
        {
            even_pos_sum += sum_digits(remainder * 2);
        }
        else
        {
            odd_pos_sum += remainder;
        }
        number = number / 10;
        idx++;
    }

    if ((even_pos_sum + odd_pos_sum) % 10 == 0)
    {
        return true;
    }
    else
    {
        return false;
    }
}

const char* check_credit_card_brand(long cc_number)
{
    // America Express 15 digitos, começam com 34 e 37 -> AMEX\n
    if ((cc_number / 10000000000000) == 34 || (cc_number / 10000000000000) == 37)
    {
        /* printf("AMEX\n"); */
        return "AMEX";
    }
    // Master Card 16 digitos, começam com 51 a 55 -> MASTERCARD\n
    else if ((cc_number / 100000000000000) == 51 || (cc_number / 100000000000000) == 52 || (cc_number / 100000000000000) == 53 || (cc_number / 100000000000000) == 54 || (cc_number / 100000000000000) == 55)
    {
        /* printf("MASTERCARD\n"); */
        return "MASTERCARD";
    }
    // Visa usa 13 e 16 Digitos, começam com 4 -> VISA\n
    else if ((cc_number / 1000000000000) == 4 || (cc_number / 1000000000000000) == 4 )
    {
        /* printf("VISA\n"); */
        return "VISA";
    }
    return "INVALID";
}


int main(void)
{
    long cc_number = get_long("CC number: ");
    bool cc_number_is_valid = check_luhn(cc_number);

    if ( cc_number_is_valid)
    {
        printf("%s\n", check_credit_card_brand(cc_number));
    }
    else
    {
        printf("INVALID\n");
    }
    return 0;
}

File: pset6/credit/credit.py
from dataclasses import dataclass
import re
import sys

from cs50 import get_int


@dataclass
class regex_in:
    """
    https://stackoverflow.com/questions/70680363/structural-pattern-matching-using-regex
    """

    string: str

    def __eq__(self, other: str | re.Pattern):
        if isinstance(other, str):
            other = re.compile(other)
        assert isinstance(other, re.Pattern)
        return other.fullmatch(self.string) is not None


def check_credit_card_brand(cc_number):
    match regex_in(str(cc_number)[:2]):
        case r"3[4|7]":
            return "AMEX"
        case r"5[1-5]":
            return "MASTERCARD"
        case r"4[0-9]":
            return "VISA"
        case _:
            return "INVALID"


def sum_digits(number):
    if number > 9:
        return number % 10 + 1
    return number


def check_luhn(cc_number):
    number = cc_number
    remainder = 0
    even_pos_sum = 0
    odd_pos_sum = 0
    idx = 1

    while number != 0:
        remainder = number % 10
        if idx % 2 == 0:
            even_pos_sum += sum_digits(remainder * 2)
        else:
            odd_pos_sum += remainder

        number = number // 10
        idx += 1

    if (even_pos_sum + odd_pos_sum) % 10 == 0:
        return True
    else:
        return False


def main():
    cc_number = get_int("CC number: ")
    cc_number_is_valid = check_luhn(cc_number)
    brand = check_credit_card_brand(cc_number)

    if cc_number_is_valid:
        print(f"{brand}")
    else:
        print(f"{brand}\n")
        return 1
    return 0


if __name__ == "__main__":
    sys.exit(main())
